test/functional/collection_tests.js

```diff
@@ -1616,6 +1620,141 @@ describe('Collection', function() {
     });
   });
 
+  it('countDocuments should return Promise that resolves when no callback passed', function(done) {
+    const configuration = this.configuration;
+    const client = new MongoClient(configuration.url(), { w: 1 });
+
+    client.connect(function(err, client) {
+      const db = client.db(configuration.db);
+      const collection = db.collection('countDoc_return_promise');
+      const docsPromise = collection.countDocuments();
+      const close = e => client.close(() => done(e));
+
+      expect(docsPromise).to.exist.and.to.be.an.instanceof(collection.s.promiseLibrary);
+
+      docsPromise
+        .then(result => expect(result).to.equal(0))
+        .then(() => close())
+        .catch(e => close(e));
+    });
+  });
+
+  it('countDocuments should not return a promise if callback given', function(done) {
+    const configuration = this.configuration;
+    const client = new MongoClient(configuration.url(), { w: 1 });
+
+    client.connect(function(err, client) {
+      const db = client.db(configuration.db);
+      const collection = db.collection('countDoc_no_promise');
+      const close = e => client.close(() => done(e));
+
+      const notPromise = collection.countDocuments({ a: 1 }, function() {
+        expect(notPromise).to.be.undefined;
+        close();
+      });
+    });
+  });
+
+  it('countDocuments should correctly call the given callback', function(done) {
+    const configuration = this.configuration;
+    const client = new MongoClient(configuration.url(), { w: 1 });
+
+    client.connect(function(err, client) {
+      const db = client.db(configuration.db);
+      const collection = db.collection('countDoc_callback');
+      const docs = [{ a: 1 }, { a: 2 }];
+      const close = e => client.close(() => done(e));
+
+      collection.insertMany(docs).then(() =>
+        collection.countDocuments({ a: 1 }, (err, data) => {
+          expect(data).to.equal(1);
+          close(err);
+        })
+      );
+    });
+  });
+
+  describe('countDocuments with mock server', function() {
+    let server;
+
+    beforeEach(() => {
+      return mock.createServer().then(s => {
+        server = s;
+      });
+    });
+
+    afterEach(() => mock.cleanup());
+
+    function testSkipLimit(config, done) {
```

Comment:
should this still be called `testSkipLimit` if it is being used in `'countDocuments should return appropriate error if aggregation fails'`?

Reply:
good catch, my bad